crates/sage/src/scoring.rs

```diff
@@ -327,61 +329,94 @@ impl<'db> Scorer<'db> {
         }
     }
 
-    fn initial_hits(&self, query: &ProcessedSpectrum, precursor: &Precursor) -> InitialHits {
+    fn initial_hits(&self, query: &ProcessedSpectrum, precursors: &[Precursor]) -> InitialHits {
         // Sage operates on masses without protons; [M] instead of [MH+]
-        let mz = precursor.mz - PROTON;
+        let mzs = precursors
+            .iter()
+            .map(|p| p.mz - PROTON)
+            .collect::<Vec<f32>>();
 
         // Search in wide-window/DIA mode
         if self.wide_window {
-            let mut hits = (self.min_precursor_charge..=self.max_precursor_charge).fold(
-                InitialHits::default(),
-                |mut hits, precursor_charge| {
-                    let precursor_mass = mz * precursor_charge as f32;
-                    let precursor_tol = precursor
-                        .isolation_window
-                        .unwrap_or(Tolerance::Da(-2.4, 2.4))
-                        * precursor_charge as f32;
-                    hits +=
-                        self.matched_peaks(query, precursor_mass, precursor_charge, precursor_tol);
-                    hits
-                },
-            );
+            let mut hits = InitialHits::default();
+            for (i, (precursor, mz)) in precursors.iter().zip(mzs).enumerate() {
+                let local_hits = (self.min_precursor_charge..=self.max_precursor_charge).fold(
+                    InitialHits::default(),
+                    |mut local_hits, precursor_charge| {
+                        let precursor_mass = mz * precursor_charge as f32;
+                        let precursor_tol = precursor
+                            .isolation_window
+                            .unwrap_or(Tolerance::Da(-2.4, 2.4))
+                            * precursor_charge as f32;
+                        let mut matched_peaks = self.matched_peaks(
+                            query,
+                            precursor_mass,
+                            precursor_charge,
+                            precursor_tol,
+                        );
+                        matched_peaks.preliminary.iter_mut().for_each(|score| {
+                            score.precursor_index = i;
```

Comment:
This could be set in the `matched_peaks` function to avoid unnecessary iteration over candidates that are never matched.